feat(agents): Add initial rules for dogfooding

sayansisodiya opened the pull request:
### What
Add initial set of rules from [sandbox repo](https://github.com/sourcegraph/review-agent-sandbox/tree/main/.sourcegraph) to `sg/cody` for dogfooding. These rules are now pulled in as context for Cody chats/edits on relevant files, and they are defined in `*.rule.md` files in `.sourcegraph/` dirs around the repo.

I put general rules in the topmost `.sourcegraph/` dir, and more specific rules at the relevant dir levels.

Also enabled rules in `settings.json` by default, so they should be used when debugging Cody locally.

### Why
We need to dogfood these rules internally. We need more info on what sort of rules are helpful, unobtrusive, and don't have too many false positives/negatives. This will guide how we create, apply, and organize rules going forward.

## Test plan
CI & tested rules locally. Ensured that rules are enabled by default when debugging Cody locally (i.e., even when user hasn't explicitly set `"cody.rules.enabled": true` in their `settings.json`).
<!-- Required. See https://docs-legacy.sourcegraph.com/dev/background-information/testing_principles. -->

dominiccooney commented on .sourcegraph/bash-style.rule.md:
````diff
@@ -0,0 +1,30 @@
+```
+
+
+Set `-eu -o pipefail` (this is generally "bash strict mode" and sets).
````
and sets...?

dominiccooney commented on .sourcegraph/bash-style.rule.md:
````diff
@@ -0,0 +1,30 @@
+
+Use `mapfile` instead of `while IFS` to read a file:
+
+```sh
````
Is this a good example? Because the "use instead" doesn't do the same thing as the "instead of". In my experience LLMs generally perform better if you avoid non-essential inconsistencies like this.

dominiccooney commented on .sourcegraph/instanceof-error.rule.md:
```diff
@@ -0,0 +1,43 @@
+---
+title: Verify instanceof errors
+description: Ensure that instanceof errors are used to check for specific error types in TypeScript
+lang: typescript
```
Will be interested to see if your TypeScript capitalization rule complains about this 😅

dominiccooney commented on .sourcegraph/instanceof-error.rule.md:
````diff
@@ -0,0 +1,43 @@
+}
+```
+
+It is acceptable to skip this rule if the error being caught is already known to be an instance of a specific error type:
````
Your previous example isn't in harmony with this rule. In a sense, you're telling the LLM that *spurious* instanceof checks are OK... do we want to be telling the LLM that?

sayansisodiya replied:
Good point, I agree it makes sense to leave out that caveat.

dominiccooney commented on .sourcegraph/no-hardcoded-strings.rule.md:
```diff
@@ -0,0 +1,27 @@
+tags: ["style"]
+---
+
+Avoid using a hardcoded string in multiple places. Instead, use a constant or variable.
```
I am anxious about this rule being applied to TypeScript, where it is common to basically use a union-of-string literals type as an enumeration, and repeating those strings to mention members of the "can't believe it is not an enum." It would be tedious to have to write constants everywhere in addition to that.

dominiccooney commented on .sourcegraph/typescript-style.rule.md:
````diff
@@ -0,0 +1,42 @@
+Use this:
+
+```typescript
+type ComponentState = {
````
It would be more convenient for the consumer, I think, to do a union with a boolean like:

```
type ComponentState = LoadingComponent | LoadedComponent | ErroredComponent

type LoadingComponent = {
  isLoading: true
}

type ErroredComponent = {
  isLoading: false,
  error: Error
}

type LoadedComponent = {
  isLoading: false,
  user: User
}
```

Or even better, `state: 'loading' | 'loaded' | 'errored'` so we can switch conveniently on it.

dominiccooney reviewed (changes requested):
Some ideas inline.

Could we have a CI where the agent runs with proposed changes, so we can review the output of the proposals in context?

sayansisodiya commented:
Thanks @dominiccooney and @karlclement for the rule feedback. I updated those rules and added a bunch more (from sg/sg and sg/review-agent-sandbox) that Karl defined for dogfooding.

olafurpg reviewed (dismissed):
I just merged the PR that adds these rules to sg/sg, and updated the agent config to enable sg/sg rules globally so we don't need to merge a copy into this repo.

olafurpg commented:
The feedback in this review is great, btw! Thank you @dominiccooney 

@sayansisodiya we should incorporate that feedback into the sg/sg rules.

sayansisodiya commented:
@olafurpg I believe I did incorporate those changes into the rules that got merged into sg/sg.

> I just merged the PR that adds these rules to sg/sg, and updated the agent config to enable sg/sg rules globally so we don't need to merge a copy into this repo.

Yea we don't need them here for the review agent, but my thinking was merging them here would allow for dogfooding "rules-in-the-editor" when working on sg/cody locally, as those rules are not synced globally like the review agent.

We can remove them later, but I think for dogfooding purposes it would be nice to have them in the editor as well.

olafurpg commented:
Fair enough. I am concerned we're making things a bit confusing by keeping multiple copies of the same rules across different repos. I would be nice if Cody could pick up global rules instead. In the meantime, keeping copies is fine. 

After a large refactoring last Friday, it should be trivial to deduplicate rules if we aren't doing it already.

sayansisodiya commented:
Sounds good, @olafurpg. Can you stamp when you get the chance? Thanks 🙏

olafurpg commented:
I think it's best for somebody actively working in the Cody repo to stamp since they'll do the dogfooding and improve the in-editor experience
